fedimint-client-rpc/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+use fedimint_mint_client::{MintClientInit, MintClientModule};
+use futures::future::{AbortHandle, Abortable};
+use futures::StreamExt;
+use serde::{Deserialize, Serialize};
+use tokio::sync::Mutex;
+use tracing::info;
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct RpcRequest {
+    pub request_id: u64,
+    #[serde(flatten)]
+    pub kind: RpcRequestKind,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum RpcRequestKind {
+    JoinFederation {
+        invite_code: String,
+        client_name: String,
+    },
+    OpenClient {
+        client_name: String,
+    },
+    CloseClient {
+        client_name: String,
+    },
+    ClientRpc {
+        client_name: String,
+        module: String,
+        method: String,
+        payload: serde_json::Value,
+    },
+    CancelRpc {
+        cancel_request_id: u64,
+    },
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcResponse {
+    pub request_id: u64,
+    #[serde(flatten)]
+    pub kind: RpcResponseKind,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum RpcResponseKind {
+    Data { data: serde_json::Value },
+    Error { error: String },
+    Aborted {},
+    End {},
+}
+
+pub trait RpcResponseHandler: MaybeSend + MaybeSync {
+    fn handle_response(&self, response: RpcResponse);
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait DatabaseFactory: MaybeSend + MaybeSync + 'static {
+    async fn create_database(&self, name: &str) -> anyhow::Result<Database>;
+}
+
+pub struct RpcGlobalState<D> {
+    clients: Mutex<HashMap<String, ClientHandleArc>>,
+    rpc_handles: std::sync::Mutex<HashMap<u64, AbortHandle>>,
+    db_factory: D,
+}
+
+pub struct HandledRpc<'a> {
+    pub task: Option<BoxFuture<'a, ()>>,
+}
+
+impl<D: DatabaseFactory> RpcGlobalState<D> {
+    pub fn new(db_factory: D) -> Self {
+        Self {
+            clients: Mutex::new(HashMap::new()),
+            rpc_handles: std::sync::Mutex::new(HashMap::new()),
+            db_factory,
+        }
+    }
+
+    async fn add_client(&self, client_name: String, client: ClientHandleArc) {
+        let mut clients = self.clients.lock().await;
+        clients.insert(client_name, client);
+    }
+
+    async fn get_client(&self, client_name: &str) -> Option<ClientHandleArc> {
+        let clients = self.clients.lock().await;
+        clients.get(client_name).cloned()
+    }
+
+    fn add_rpc_handle(&self, request_id: u64, handle: AbortHandle) {
+        let mut handles = self.rpc_handles.lock().unwrap();
+        if handles.insert(request_id, handle).is_some() {
+            tracing::error!("RPC CLIENT ERROR: request id reuse detected");
+        }
+    }
+
+    fn remove_rpc_handle(&self, request_id: u64) -> Option<AbortHandle> {
+        let mut handles = self.rpc_handles.lock().unwrap();
+        handles.remove(&request_id)
+    }
+
+    async fn handle_join_federation(
```

Comment:
I see, so the whole thing is stateful and multi-federation, so it keeps track of which federations were joined etc.?

Reply:
it doesn't keep track of the list, you need to keep track of list of joined clients somewhere else

but yes it is stateful